src/coins.cpp

```diff
@@ -659,6 +665,27 @@ bool CCoinsViewCache::GetSidechain(const uint256 & scId, CSidechain& targetScInf
     return false;
 }
 
+int64_t CCoinsViewCache::GetTopQualityCert(const uint256& scId, int epochNumber, uint256& hash) const
+{
+    LogPrint("mempool", "%s.%s():%d - sc %s, epoch %d\n", __FILE__, __func__, __LINE__, scId.ToString(), epochNumber);
+
+    int64_t topQual = base->GetTopQualityCert(scId, epochNumber, hash);
+    LogPrint("mempool", "%s.%s():%d - base: cert [%s], q=%d\n", __FILE__, __func__, __LINE__, hash.ToString(), topQual);
+
+    CSidechain targetInfo;
+    if (GetSidechain(scId, targetInfo)                             &&
+        !targetInfo.lastCertificateHash.IsNull()                   &&
+        epochNumber == targetInfo.lastEpochReferencedByCertificate &&
```

Comment:
I wonder if the check on lastEpochReferencesByCertificate is of any value.
Can't `GetTopQualityCert` return the highest quality for all certificates received for a given sidechain?
With the check on epochNumber, a GetTopQualityCert call after the sidechain is expired would return -1, I believe. Why you need this behaviour?